### PR TITLE
GH-37753: [C++][Gandiva] Add external function registry support

### LINKED ISSUE #37753: [C++][Gandiva] Support external function registry
### Describe the enhancement requested

# Description
Our team has been leveraging Gandiva in our projects, and its performance and capabilities have been commendable. However, we've identified a constraint concerning the registration of functions. At present, Gandiva necessitates that functions be registered directly within its codebase. This method, while functional, is not the most user-friendly and presents hurdles for those aiming to incorporate third-party functions. Direct modifications to Gandiva's source code for such integrations can inadvertently introduce maintenance challenges and potential versioning conflicts down the line.

# Proposal
To address this limitation, I propose the introduction of an external function registry mechanism in Gandiva. This would allow users and developers to register and integrate custom functions without directly modifying Gandiva's core source code.
Two major changes are proposed for supporting this capability:
* add a new API `AddFunction(NativeFunction native_function)` for `FunctionRegistry`, where the given parameter `native_function` stores the external function metadata, so that developers can register external functions by calling this API. The external function metadata discovery responsibility is outside the scope of this proposal and developers can come up with her own 
* add a new class for storing external functions' LLVM IR buffers (likely a singleton class), and merge the external IRs with the built-in function IR into the LLVM module, so that third party pre-compiled functions can be integrated via LLVM bitcode

~~I've came up with a PR (https://github.com/apache/arrow/pull/37787) with more details, including the JSON schema and a minimum external C++ external function registered via this approach to demonstrate/test how it works.~~
The latest PR for implementing this proposal is here (https://github.com/apache/arrow/pull/38116)

# Benefits
* Flexibility: This feature would grant users the autonomy to effortlessly integrate third-party functions or bespoke logic, eliminating the need to navigate the core codebase.
* Maintainability: Segregating external functions ensures that Gandiva's primary code remains streamlined, facilitating easier updates and maintenance.
* Polyglot programming: The proposed registry could pave the way for function authoring beyond C++, potentially embracing languages like C, Rust, or Zig. These could then be integrated via a standard C API or even at the LLVM IR level.
* Community Engagement: Encouraging external function registration could bolster community participation. This would enable developers to not only contribute but also disseminate their unique functions. Consequently, specialized functions could be curated and shared more widely.

While the intricate design specifics are still under consideration, I'm keen to understand the community's perspective on this proposal. Would an external function registry align with Gandiva's future roadmap? Your insights would be invaluable. Thank you.

### Component(s)

C++ - Gandiva

UPDATE:
* 2023-10-07: submit a new PR (https://github.com/apache/arrow/pull/38116) and close the original PR (https://github.com/apache/arrow/pull/37787) after discussion in the dev mailing list (Related mailing list discussion: https://lists.apache.org/thread/lm4sbw61w9cl7fsmo7tz3gvkq0ox6rod)
* 2023-09-28: according to mailing list's feedback, revise the proposal by removing the JSON/directory metadata population logic out of this proposal
* 2023-09-25: I added some more description into the proposal section

### PR DESCRIPTION
**Pull request by niyue**
Add external function registry support to Gandiva. JSON based function registry can be used for describing the function metadata, and LLVM bitcode can be automatically loaded as pre-compiled external functions.

### Rationale for this change
This PR tries to enhance Gandiva by supporting external function registry, so that developers can author third party functions without modifying Gandiva's core codebase. See https://github.com/apache/arrow/issues/37753 for more details.

### What changes are included in this PR?
This PR implements basic support of external function registry. It primarily implements two parts to make it work:
1) define a JSON based function registry format to describe the function metadata, and support loading and parsing such JSON based function registry files from some directory
2) automatically load and parse LLVM bitcode from files, so that third party pre-compiled functions can be integrated via LLVM bitcode

The overall flow looks like this:
external-func-registry

### Are these changes tested?
Some unit tests are added to verify this enhancement

### Are there any user-facing changes?
No change to the existing behavior. But some new ways to interfacing the library are added in this PR.

* Closes: #37753

# Notes
* Performance
  * since the function registry is loaded once and stored as a static map internally, there shouldn't too much performance impact typically by externally loading the JSON function registry files. 
  * If external function bitcodes are added, loading them from disk may take some additional time (but it won't affect anything if no external function bitcode are added). I am not sure if we should cache such parsed bitcode in memory for better performance, please shed some light on this. Thanks so much.
* Integration with other programming languages via LLVM IR/bitcode
  * So far I only added an external C++ function in the codebase for unit testing purpose. Rust based function is possible but I gave it a try and found another issue (Rust has std lib which needs to be processed in different approach), I will do some exploration for other languages such as zig later.
* Non pre-compiled functions, may require some different approach to get the function pointer, and we may discuss and work on it in a separated PR later. I am currently thinking using approach like loading shared libraries during runtime, and find the corresponding function pointer symbols from the shared library, but I don't do any experiment yet so I am not sure if it works this way. Any comment on how this should be done is appreciated.
* Certain complex data types, such as map, struct, are not allowed to be used in registry yet, as far as I know, Gandiva doesn't support using these types as input/output for function either, but the registry format could be extended to support them in the future if needed
* JSON schema for the external registry format
```JSON
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "functions": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "aliases": {
            "type": "array",
            "items": {
              "type": "string"
            }
          },
          "param_types": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "type": {
                  "type": "string",
                  "enum": ["null", "boolean", "uint8", "int8", "uint16", "int16", "uint32", "int32", "uint64", "int64", "float16", "float32", "float64", "utf8", "large_utf8", "binary", "large_binary", "date32", "date64", "day_time_interval", "month_interval", "timestamp", "decimal", "decimal128", "decimal256", "list"]
                },
                "unit": {
                  "type": "string",
                  "enum": ["second", "milli", "micro", "nano"]
                },
                "precision": {
                  "type": "integer"
                },
                "scale": {
                  "type": "integer"
                },
                "value_type": {
                  "$ref": "#/properties/functions/items/properties/param_types/items"
                }
              },
              "required": ["type"]
            }
          },
          "return_type": {
            "$ref": "#/properties/functions/items/properties/param_types/items"
          },
          "result_nullable": {
            "type": "string",
            "enum": ["ifnull", "never", "internal"]
          },
          "pc_name": {
            "type": "string"
          },
          "needs_context": {
            "type": "boolean"
          },
          "needs_function_holder": {
            "type": "boolean"
          },
          "can_return_errors": {
            "type": "boolean"
          }
        },
        "required": ["name"]
      }
    }
  },
  "required": ["functions"]
}
```

### CODE REVIEW

File: cpp/src/gandiva/function_registry_external.cc [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+#include <gandiva/function_registry_external.h>
+#include <rapidjson/document.h>
+#include <filesystem>
+#include <fstream>
+#include <vector>
+
+namespace gandiva {
+namespace rj = rapidjson;
+
+class JsonRegistryParser {
```
Comment (niyue) [inline cpp/src/gandiva/function_registry_external.cc:28]: A JSON registry parser is implemented to parse the function metadata from the JSON content.

**Comment by kou** [comment]
(Please mention me when this is ready for review.)

**Comment by niyue** [comment]
@kou this PR is ready for review now. Thanks for the help.
There are still 4 CI checks failing. But they seemed to be a failure about `test_python_version`, which is probably irrelevant with this PR.

**Comment by kou** [comment]
Could you rebase on the `Dev / Source Release and Merge Script` failures?
Could you check the failure on AppVeyor?

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48091110#L2859

```text
[ RUN      ] TestFunctionRegistry.LookupMultipleFuncs
C:/projects/arrow/cpp/src/gandiva/function_registry_test.cc(105): error: Expected: (say_hello_func) != (nullptr), actual: NULL vs (nullptr)
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
Stack trace:
[  FAILED  ] TestFunctionRegistry.LookupMultipleFuncs (9 ms)
[ RUN      ] TestFunctionRegistry.LookupExternalFuncs
C:/projects/arrow/cpp/src/gandiva/function_registry_test.cc(125): error: Expected: (func) != (nullptr), actual: NULL vs (nullptr)
unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
Stack trace:
[  FAILED  ] TestFunctionRegistry.LookupExternalFuncs (12 ms)
[----------] 5 tests from TestFunctionRegistry (41 ms total)
[----------] 6 tests from TestExternalFunctionRegistry
```

**Comment by niyue** [comment]
@kou I've rebased to the latest commit (772a01c08), but the CI ran into another issue (https://github.com/apache/arrow/actions/runs/6290902281/job/17078761957?pr=37787), which seems to be related with LLVM 17 (https://github.com/apache/arrow/issues/37834)? Do I have to fix that or is there any specific commit I should rebase onto?

> Could you check the failure on AppVeyor?
> https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48091110#L2859

Do you know what the `AppVeyor` check in CI is? Because I cannot reproduce this issue locally (A M1 MacBook), and several CI checks passed as well. I would like to know what the `AppVeyor` check in CI is and how it is different compared to other C++ checks in CI to see if I can figure out why they failed in that environment.

**Comment by kou** [comment]
> @kou I've rebased to the latest commit ([772a01c](https://github.com/apache/arrow/commit/772a01c080ad57eb11e9323f5347472b769d45de)), but the CI ran into another issue (https://github.com/apache/arrow/actions/runs/6290902281/job/17078761957?pr=37787), which seems to be related with LLVM 17 (#37834)? Do I have to fix that or is there any specific commit I should rebase onto?

Oh. We can ignore the failures. We can work on it in a separated PR.

> > Could you check the failure on AppVeyor?
> > https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48091110#L2859
> 
> Do you know what the `AppVeyor` check in CI is? Because I cannot reproduce this issue locally (A M1 MacBook), and several CI checks passed as well. I would like to know what the `AppVeyor` check in CI is and how it is different compared to other C++ checks in CI to see if I can figure out why they failed in that environment. Thanks

AppVeyor uses Windows and Visual C++. The failure may be related to Windows and/or Visual C++.

**Review by kou — COMMENTED** [review]
Could you rebase on main after #37850 is merged and format `*.cmake` files by `archery lint --cmake-format --fix`?
See also: https://arrow.apache.org/docs/dev/developers/cpp/development.html#code-style-linting-and-ci

Could you add the JSON schema too?

Could you add document for this feature? Or we can defer it to a follow-up PR by creating a new issue for it.

Could you send an e-mail to `dev@arrow.apache.org` https://lists.apache.org/list.html?dev@arrow.apache.org https://arrow.apache.org/community/#mailing-lists with `[DISCUSS][Gandiva] ...` subject to get more attention for this specification?

**Comment by niyue** [comment]
> AppVeyor uses Windows and Visual C++. The failure may be related to Windows and/or Visual C++.

Previously, I made the PR to pass several Windows related checks, such as "C++ / AMD64 Windows MinGW CLANG64 C++ " and "C++ / AMD64 Windows MinGW MINGW64 C++ " (but now they are failing again due to the LLVM 17 issue). Do you know what are the difference between them and the `AppVeyor` CI check? Considering both of them are Windows environments, do they differ in compiler primarily (clang vs. Visual C++ compiler)?

**Comment by niyue** [comment]
> Could you rebase on main after #37850 is merged and format `*.cmake` files by `archery lint --cmake-format --fix`? See also: https://arrow.apache.org/docs/dev/developers/cpp/development.html#code-style-linting-and-ci

Sure. I will wait for the #37850 to be merged.

> Could you add the JSON schema too?
> Could you add document for this feature? Or we can defer it to a follow-up PR by creating a new issue for it.
Do you have any recommended folder(s) to add:
1) JSON schema
2) the corresponding doc

There are some existing folders, but none of them seem to be a good fit for either the JSON schema or doc (I don't find any doc related directory under the `src/gandiva` directory)

> Could you send an e-mail to `dev@arrow.apache.org` https://lists.apache.org/list.html?dev@arrow.apache.org https://arrow.apache.org/community/#mailing-lists with `[DISCUSS][Gandiva] ...` subject to get more attention for this specification?
Sure. I sent an email to the dev mailing list a while ago, and I will collect some feedback and see how I should proceed.

**Comment by kou** [comment]
> Do you know what are the difference between them and the `AppVeyor` CI check?

They use GCC/Clang (built on MSYS2 that is an UNIX emulation layer on Windows).
AppVeyor use Visual C++.

> Considering both of them are Windows environments, do they differ in compiler primarily (clang vs. Visual C++ compiler)?

Yes.

> Do you have any recommended folder(s) to add:
>
> 1. JSON schema

How about https://github.com/apache/arrow/tree/main/format (`format/gandiva/`)?

> 2. the corresponding doc

How about https://github.com/apache/arrow/tree/main/docs/source/format (`docs/source/format/Gandiva.rst`)?

> Sure. I sent an email to the dev mailing list a while ago, and I will collect some feedback and see how I should proceed.

Thanks!

The thread: https://lists.apache.org/thread/lm4sbw61w9cl7fsmo7tz3gvkq0ox6rod

**Comment by MehdiChinoune** [comment]
Are you aware that the code still includes a header file that doesn't exist in LLVM-17?

**Comment by niyue** [comment]
After discussing this proposal in mailing list, I will close this PR in favor of https://github.com/apache/arrow/pull/38116 Thank you all for the help!

**Pull request closed without merging**